test/controllers/journeys/income/AnyOtherIncomeControllerSpec.scala

```diff
@@ -127,40 +139,131 @@ class AnyOtherIncomeControllerSpec extends SpecBase with MockitoSugar {
 
     "onSubmit" - {
 
-      "must redirect to the next page when valid data is submitted" in {
+      "when a user answer 'Yes' is submitted, must redirect to the Other Income Amount page" - {
+        "when journey is ACCRUAL accounting type" in {
 
-        val mockSessionRepository = mock[SessionRepository]
+          val userAnswer = true
 
-        when(mockSessionRepository.set(any())) thenReturn Future.successful(true)
+          val application =
+            applicationBuilder(userAnswers = Some(emptyUserAnswers))
+              .overrides(
+                bind[IncomeNavigator].toInstance(new FakeIncomeNavigator(answerIsTrueOnwardRoute)),
+                bind[SelfEmploymentService].toInstance(mockService),
+                bind[SessionRepository].toInstance(mockSessionRepository)
+              )
+              .build()
 
-        val application =
-          applicationBuilder(userAnswers = Some(emptyUserAnswers))
-            .overrides(
-              bind[Navigator].toInstance(new FakeNavigator(onwardRoute)),
-              bind[SessionRepository].toInstance(mockSessionRepository)
-            )
-            .build()
+          running(application) {
+            when(mockSessionRepository.set(any())) thenReturn Future.successful(true)
+            when(mockService.getAccountingType(any, meq(businessId), any)(any)) thenReturn Future(Right("ACCRUAL"))
 
-        running(application) {
-          val request = FakeRequest(POST, AnyOtherIncomeController.onSubmit(taxYear, NormalMode).url)
-            .withFormUrlEncodedBody(("value", "true"))
+            val request = FakeRequest(POST, AnyOtherIncomeController.onSubmit(taxYear, businessId, NormalMode).url)
+              .withFormUrlEncodedBody(("value", userAnswer.toString))
 
-          val result = route(application, request).value
+            val result = route(application, request).value
 
-          status(result) mustEqual SEE_OTHER
-          redirectLocation(result).value mustEqual onwardRoute.url
+            status(result) mustEqual SEE_OTHER
+            redirectLocation(result).value mustEqual otherIncomeAmountCall.url
+          }
+        }
```

Comment:
I think the tests you've written here (for checking that a certain redirection happens based upon what you've set the user answers to and what the accounting type is) are not actually testing what you'd hope them to. 

The reason for this is that the responsibility of redirection based upon user answers and the accounting type in the controller is delegated to the navigator. In the tests you've written, we are essentially mocking our navigator, and so regardless of what we set our user answers or the accounting type to, this has no effect. 

To highlight what I mean, if in this test we change the user answers to false and the accounting type to "CASH", then the tests still pass. I think what you are trying to test is the integration of the controller with the navigator, which is are fair tests to do (maybe we should be doing this???), however these tests should be unit tests. 

I think for our unit tests, we could do the following:

- Test that when `selfEmploymentService.getAccountingType` returns Left(x) (which we can achieve by mocking like you have already done), that we redirect to `JourneyRecoveryController`

- Test that when it returns an accounting type, and our form data does not match the format of the form provider (which we already have tests for), we pass in that form with errors to the view

- Test that when it returns an accounting type, and our form data matches the form provider format, that user answers are set correctly (can be done without mocking) and that the session repository has successfully set our answers (done via mocking), that we are passing those updated user answers and the accounting type to the navigator (this can be done by setting the expectations on the params during mocking in Mockito, i.e. `(eq(updatedUserAnswers))` ) and we redirect to it's return value.

Hope this makes sense. Happy to discuss it if i have misunderstood.

Note - I think this comment also applies to the other controller tests where you've done this sort of testing.

Reply:
We will discuss this with the team to decide what's best going forwards